Add SolarPlantsBrazil dataset for photovoltaic panel detection (semantic segmentation)

FederCO23 opened the pull request:
This PR adds the **SolarPlantsBrazil** dataset to TorchGeo.

SolarPlantsBrazil is a geospatial dataset curated for **binary semantic segmentation**, focused on detecting **photovoltaic (PV) power stations** in satellite imagery. It includes **272 manually annotated 256×256 GeoTIFF image patches** from various regions of Brazil. Each image patch comes with:
- A **4-band image** (RGB + NIR, float32)
- A **binary mask** indicating solar panel regions

### Changes introduced:
- Added `SolarPlantsBrazil` dataset class under `torchgeo.datasets`
- Added `SolarPlantsBrazilDataModule` datamodule class under `torchgeo.datamodules`
- Integrated automatic download from Hugging Face: [FederCO23/solar-plants-brazil](https://huggingface.co/datasets/FederCO23/solar-plants-brazil)
- Parsed RGB + NIR float32 GeoTIFFs and binary mask labels
- Implemented a `plot()` method using RGB bands for sample visualization
- Registered dataset in `__init__.py` and `non_geo_datasets.csv`
- Added full unit test coverage under `tests/datasets/test_solar_plants_brazil.py`

### Example output from `plot()`:

![Sample_from_Solar_Plants_Brazil](https://github.com/user-attachments/assets/a0122498-85d2-4ea0-b0dc-3f45ffc26094)



### References:
- [ Hugging Face Dataset](https://huggingface.co/datasets/FederCO23/solar-plants-brazil)
- [ Project Repository](https://github.com/FederCO23/UCSD_MLBootcamp_Capstone)

FederCO23 commented:
@microsoft-github-policy-service agree



On Fri, May 23, 2025 at 9:02 PM microsoft-github-policy-service[bot] <
***@***.***> wrote:

> *microsoft-github-policy-service[bot]* left a comment
> (microsoft/torchgeo#2797)
> <https://github.com/microsoft/torchgeo/pull/2797#issuecomment-2906058617>
>
> @FederCO23 <https://github.com/FederCO23> please read the following
> Contributor License Agreement(CLA). If you agree with the CLA, please reply
> with the following information.
>
> @microsoft-github-policy-service agree [company="{your company}"]
>
> Options:
>
>    - (default - no company specified) I have sole ownership of
>    intellectual property rights to my Submissions and I am not making
>    Submissions in the course of work for my employer.
>
> @microsoft-github-policy-service agree
>
>
>    - (when company given) I am making Submissions in the course of work
>    for my employer (or my employer has intellectual property rights in my
>    Submissions by contract or applicable law). I have permission from my
>    employer to make Submissions and enter into this Agreement on behalf of my
>    employer. By signing below, the defined term “You” includes me and my
>    employer.
>
> @microsoft-github-policy-service agree company="Microsoft"
>
> Contributor License Agreement Contribution License Agreement
>
> This Contribution License Agreement (*“Agreement”*) is agreed to by the
> party signing below (*“You”*),
> and conveys certain license rights to Microsoft Corporation and its
> affiliates (“Microsoft”) for Your
> contributions to Microsoft open source projects. This Agreement is
> effective as of the latest signature
> date below.
>
>    1. *Definitions*.
>    *“Code”* means the computer software code, whether in human-readable
>    or machine-executable form,
>    that is delivered by You to Microsoft under this Agreement.
>    *“Project”* means any of the projects owned or managed by Microsoft
>    and offered under a license
>    approved by the Open Source Initiative (www.opensource.org).
>    *“Submit”* is the act of uploading, submitting, transmitting, or
>    distributing code or other content to any
>    Project, including but not limited to communication on electronic
>    mailing lists, source code control
>    systems, and issue tracking systems that are managed by, or on behalf
>    of, the Project for the purpose of
>    discussing and improving that Project, but excluding communication
>    that is conspicuously marked or
>    otherwise designated in writing by You as “Not a Submission.”
>    *“Submission”* means the Code and any other copyrightable material
>    Submitted by You, including any
>    associated comments and documentation.
>    2. *Your Submission*. You must agree to the terms of this Agreement
>    before making a Submission to any
>    Project. This Agreement covers any and all Submissions that You, now
>    or in the future (except as
>    described in Section 4 below), Submit to any Project.
>    3. *Originality of Work*. You represent that each of Your Submissions
>    is entirely Your original work.
>    Should You wish to Submit materials that are not Your original work,
>    You may Submit them separately
>    to the Project if You (a) retain all copyright and license information
>    that was in the materials as You
>    received them, (b) in the description accompanying Your Submission,
>    include the phrase “Submission
>    containing materials of a third party:” followed by the names of the
>    third party and any licenses or other
>    restrictions of which You are aware, and (c) follow any other
>    instructions in the Project’s written
>    guidelines concerning Submissions.
>    4. *Your Employer*. References to “employer” in this Agreement include
>    Your employer or anyone else
>    for whom You are acting in making Your Submission, e.g. as a
>    contractor, vendor, or agent. If Your
>    Submission is made in the course of Your work for an employer or Your
>    employer has intellectual
>    property rights in Your Submission by contract or applicable law, You
>    must secure permission from Your
>    employer to make the Submission before signing this Agreement. In that
>    case, the term “You” in this
>    Agreement will refer to You and the employer collectively. If You
>    change employers in the future and
>    desire to Submit additional Submissions for the new employer, then You
>    agree to sign a new Agreement
>    and secure permission from the new employer before Submitting those
>    Submissions.
>    5. *Licenses*.
>
>
>    - *Copyright License*. You grant Microsoft, and those who receive the
>    Submission directly or
>    indirectly from Microsoft, a perpetual, worldwide, non-exclusive,
>    royalty-free, irrevocable license in the
>    Submission to reproduce, prepare derivative works of, publicly
>    display, publicly perform, and distribute
>    the Submission and such derivative works, and to sublicense any or all
>    of the foregoing rights to third
>    parties.
>    - *Patent License*. You grant Microsoft, and those who receive the
>    Submission directly or
>    indirectly from Microsoft, a perpetual, worldwide, non-exclusive,
>    royalty-free, irrevocable license under
>    Your patent claims that are necessarily infringed by the Submission or
>    the combination of the
>    Submission with the Project to which it was Submitted to make, have
>    made, use, offer to sell, sell and
>    import or otherwise dispose of the Submission alone or with the
>    Project.
>    - *Other Rights Reserved*. Each party reserves all rights not
>    expressly granted in this Agreement.
>    No additional licenses or rights whatsoever (including, without
>    limitation, any implied licenses) are
>    granted by implication, exhaustion, estoppel or otherwise.
>
>
>    6. *Representations and Warranties*. You represent that You are
>    legally entitled to grant the above
>    licenses. You represent that each of Your Submissions is entirely Your
>    original work (except as You may
>    have disclosed under Section 3). You represent that You have secured
>    permission from Your employer to
>    make the Submission in cases where Your Submission is made in the
>    course of Your work for Your
>    employer or Your employer has intellectual property rights in Your
>    Submission by contract or applicable
>    law. If You are signing this Agreement on behalf of Your employer, You
>    represent and warrant that You
>    have the necessary authority to bind the listed employer to the
>    obligations contained in this Agreement.
>    You are not expected to provide support for Your Submission, unless
>    You choose to do so. UNLESS
>    REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING, AND EXCEPT FOR THE
>    WARRANTIES
>    EXPRESSLY STATED IN SECTIONS 3, 4, AND 6, THE SUBMISSION PROVIDED
>    UNDER THIS AGREEMENT IS
>    PROVIDED WITHOUT WARRANTY OF ANY KIND, INCLUDING, BUT NOT LIMITED TO,
>    ANY WARRANTY OF
>    NONINFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
>    7. *Notice to Microsoft*. You agree to notify Microsoft in writing of
>    any facts or circumstances of which
>    You later become aware that would make Your representations in this
>    Agreement inaccurate in any
>    respect.
>    8. *Information about Submissions*. You agree that contributions to
>    Projects and information about
>    contributions may be maintained indefinitely and disclosed publicly,
>    including Your name and other
>    information that You submit with Your Submission.
>    9. *Governing Law/Jurisdiction*. This Agreement is governed by the
>    laws of the State of Washington, and
>    the parties consent to exclusive jurisdiction and venue in the federal
>    courts sitting in King County,
>    Washington, unless no federal subject matter jurisdiction exists, in
>    which case the parties consent to
>    exclusive jurisdiction and venue in the Superior Court of King County,
>    Washington. The parties waive all
>    defenses of lack of personal jurisdiction and forum non-conveniens.
>    10. *Entire Agreement/Assignment*. This Agreement is the entire
>    agreement between the parties, and
>    supersedes any and all prior agreements, understandings or
>    communications, written or oral, between
>    the parties relating to the subject matter hereof. This Agreement may
>    be assigned by Microsoft.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/microsoft/torchgeo/pull/2797#issuecomment-2906058617>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BIQQR3ODLK34VV6GRDRWWKD276ZJFAVCNFSM6AAAAAB5Z3M4TOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDSMBWGA2TQNRRG4>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

FederCO23 commented:
@microsoft-github-policy-service agree


On Fri, May 23, 2025 at 9:02 PM microsoft-github-policy-service[bot] <
***@***.***> wrote:

> *microsoft-github-policy-service[bot]* left a comment
> (microsoft/torchgeo#2797)
> <https://github.com/microsoft/torchgeo/pull/2797#issuecomment-2906058617>
>
> @FederCO23 <https://github.com/FederCO23> please read the following
> Contributor License Agreement(CLA). If you agree with the CLA, please reply
> with the following information.
>
> @microsoft-github-policy-service agree [company="{your company}"]
>
> Options:
>
>    - (default - no company specified) I have sole ownership of
>    intellectual property rights to my Submissions and I am not making
>    Submissions in the course of work for my employer.
>
> @microsoft-github-policy-service agree
>
>
>    - (when company given) I am making Submissions in the course of work
>    for my employer (or my employer has intellectual property rights in my
>    Submissions by contract or applicable law). I have permission from my
>    employer to make Submissions and enter into this Agreement on behalf of my
>    employer. By signing below, the defined term “You” includes me and my
>    employer.
>
> @microsoft-github-policy-service agree company="Microsoft"
>
> Contributor License Agreement Contribution License Agreement
>
> This Contribution License Agreement (*“Agreement”*) is agreed to by the
> party signing below (*“You”*),
> and conveys certain license rights to Microsoft Corporation and its
> affiliates (“Microsoft”) for Your
> contributions to Microsoft open source projects. This Agreement is
> effective as of the latest signature
> date below.
>
>    1. *Definitions*.
>    *“Code”* means the computer software code, whether in human-readable
>    or machine-executable form,
>    that is delivered by You to Microsoft under this Agreement.
>    *“Project”* means any of the projects owned or managed by Microsoft
>    and offered under a license
>    approved by the Open Source Initiative (www.opensource.org).
>    *“Submit”* is the act of uploading, submitting, transmitting, or
>    distributing code or other content to any
>    Project, including but not limited to communication on electronic
>    mailing lists, source code control
>    systems, and issue tracking systems that are managed by, or on behalf
>    of, the Project for the purpose of
>    discussing and improving that Project, but excluding communication
>    that is conspicuously marked or
>    otherwise designated in writing by You as “Not a Submission.”
>    *“Submission”* means the Code and any other copyrightable material
>    Submitted by You, including any
>    associated comments and documentation.
>    2. *Your Submission*. You must agree to the terms of this Agreement
>    before making a Submission to any
>    Project. This Agreement covers any and all Submissions that You, now
>    or in the future (except as
>    described in Section 4 below), Submit to any Project.
>    3. *Originality of Work*. You represent that each of Your Submissions
>    is entirely Your original work.
>    Should You wish to Submit materials that are not Your original work,
>    You may Submit them separately
>    to the Project if You (a) retain all copyright and license information
>    that was in the materials as You
>    received them, (b) in the description accompanying Your Submission,
>    include the phrase “Submission
>    containing materials of a third party:” followed by the names of the
>    third party and any licenses or other
>    restrictions of which You are aware, and (c) follow any other
>    instructions in the Project’s written
>    guidelines concerning Submissions.
>    4. *Your Employer*. References to “employer” in this Agreement include
>    Your employer or anyone else
>    for whom You are acting in making Your Submission, e.g. as a
>    contractor, vendor, or agent. If Your
>    Submission is made in the course of Your work for an employer or Your
>    employer has intellectual
>    property rights in Your Submission by contract or applicable law, You
>    must secure permission from Your
>    employer to make the Submission before signing this Agreement. In that
>    case, the term “You” in this
>    Agreement will refer to You and the employer collectively. If You
>    change employers in the future and
>    desire to Submit additional Submissions for the new employer, then You
>    agree to sign a new Agreement
>    and secure permission from the new employer before Submitting those
>    Submissions.
>    5. *Licenses*.
>
>
>    - *Copyright License*. You grant Microsoft, and those who receive the
>    Submission directly or
>    indirectly from Microsoft, a perpetual, worldwide, non-exclusive,
>    royalty-free, irrevocable license in the
>    Submission to reproduce, prepare derivative works of, publicly
>    display, publicly perform, and distribute
>    the Submission and such derivative works, and to sublicense any or all
>    of the foregoing rights to third
>    parties.
>    - *Patent License*. You grant Microsoft, and those who receive the
>    Submission directly or
>    indirectly from Microsoft, a perpetual, worldwide, non-exclusive,
>    royalty-free, irrevocable license under
>    Your patent claims that are necessarily infringed by the Submission or
>    the combination of the
>    Submission with the Project to which it was Submitted to make, have
>    made, use, offer to sell, sell and
>    import or otherwise dispose of the Submission alone or with the
>    Project.
>    - *Other Rights Reserved*. Each party reserves all rights not
>    expressly granted in this Agreement.
>    No additional licenses or rights whatsoever (including, without
>    limitation, any implied licenses) are
>    granted by implication, exhaustion, estoppel or otherwise.
>
>
>    6. *Representations and Warranties*. You represent that You are
>    legally entitled to grant the above
>    licenses. You represent that each of Your Submissions is entirely Your
>    original work (except as You may
>    have disclosed under Section 3). You represent that You have secured
>    permission from Your employer to
>    make the Submission in cases where Your Submission is made in the
>    course of Your work for Your
>    employer or Your employer has intellectual property rights in Your
>    Submission by contract or applicable
>    law. If You are signing this Agreement on behalf of Your employer, You
>    represent and warrant that You
>    have the necessary authority to bind the listed employer to the
>    obligations contained in this Agreement.
>    You are not expected to provide support for Your Submission, unless
>    You choose to do so. UNLESS
>    REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING, AND EXCEPT FOR THE
>    WARRANTIES
>    EXPRESSLY STATED IN SECTIONS 3, 4, AND 6, THE SUBMISSION PROVIDED
>    UNDER THIS AGREEMENT IS
>    PROVIDED WITHOUT WARRANTY OF ANY KIND, INCLUDING, BUT NOT LIMITED TO,
>    ANY WARRANTY OF
>    NONINFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
>    7. *Notice to Microsoft*. You agree to notify Microsoft in writing of
>    any facts or circumstances of which
>    You later become aware that would make Your representations in this
>    Agreement inaccurate in any
>    respect.
>    8. *Information about Submissions*. You agree that contributions to
>    Projects and information about
>    contributions may be maintained indefinitely and disclosed publicly,
>    including Your name and other
>    information that You submit with Your Submission.
>    9. *Governing Law/Jurisdiction*. This Agreement is governed by the
>    laws of the State of Washington, and
>    the parties consent to exclusive jurisdiction and venue in the federal
>    courts sitting in King County,
>    Washington, unless no federal subject matter jurisdiction exists, in
>    which case the parties consent to
>    exclusive jurisdiction and venue in the Superior Court of King County,
>    Washington. The parties waive all
>    defenses of lack of personal jurisdiction and forum non-conveniens.
>    10. *Entire Agreement/Assignment*. This Agreement is the entire
>    agreement between the parties, and
>    supersedes any and all prior agreements, understandings or
>    communications, written or oral, between
>    the parties relating to the subject matter hereof. This Agreement may
>    be assigned by Microsoft.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/microsoft/torchgeo/pull/2797#issuecomment-2906058617>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BIQQR3ODLK34VV6GRDRWWKD276ZJFAVCNFSM6AAAAAB5Z3M4TOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDSMBWGA2TQNRRG4>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

robmarkcole commented on torchgeo/datasets/solar_plants_brazil.py:
```diff
@@ -0,0 +1,189 @@
+    """
+
+    url = 'https://huggingface.co/datasets/FederCO23/solar-plants-brazil'
+    bands = ('Red', 'Green', 'Blue', 'NIR')
```
I wonder if we could start using enums - [e.g. like terratorch ](https://github.com/IBM/terratorch/blob/main/terratorch/datasets/utils.py
)

adamjstewart replied:
I've thought about it, but there's a lot more info I would like to add (min/max/mean wavelength, bandwidth, etc.). I think a formal class would be better. But that's for another day, we would have to update all existing datasets first.

robmarkcole commented:
@FederCO23 care to add a datamodule too?

FederCO23 commented on torchgeo/datasets/solar_plants_brazil.py:
```diff
@@ -0,0 +1,189 @@
+
+    If you use this dataset, please cite or reference the project repository.
+
+    .. versionadded:: 0.6
```
done 👍

FederCO23 commented on tests/datasets/test_solar_plants_brazil.py:
```diff
@@ -0,0 +1,138 @@
+import os
+import tempfile
+import warnings
+from unittest.mock import patch
```
Replaced unittest.mock.patch with pytest.MonkeyPatch

FederCO23 commented on tests/datasets/test_solar_plants_brazil.py:
```diff
@@ -0,0 +1,138 @@
+            dst.write(mask, 1)
+
+
+def test_solar_plants_brazil_getitem_and_len() -> None:
```
Refactored tests into a class (TestSolarPlantsBrazil) with individual method.

FederCO23 commented on torchgeo/datasets/solar_plants_brazil.py:
```diff
@@ -0,0 +1,189 @@
+import numpy as np
+import rasterio
+import torch
+from huggingface_hub import snapshot_download
```
Ok. Removed huggingface_hub; now using `download_and_extract_archive`

FederCO23 commented on torchgeo/datasets/solar_plants_brazil.py:
```diff
@@ -0,0 +1,189 @@
+    url = 'https://huggingface.co/datasets/FederCO23/solar-plants-brazil'
+    bands = ('Red', 'Green', 'Blue', 'NIR')
+
+    citation = """\
```
Ok. Removed citation block from the dataset class

FederCO23 commented on torchgeo/datasets/solar_plants_brazil.py:
```diff
@@ -0,0 +1,189 @@
+        )
+
+        if len(self.image_paths) == 0:
+            raise FileNotFoundError(
```
Ok. Replaced FileNotFoundError with DatasetNotFoundError, and documented it in the docstring.

robmarkcole commented:
Test failures same as https://github.com/microsoft/torchgeo/pull/2707#issuecomment-2908777476

robmarkcole commented:
@FederCO23 hit the `Update branch` button as the failing test is fixed on main

FederCO23 commented:
> @FederCO23 hit the `Update branch` button as the failing test is fixed on main

@robmarkcole, I’ve updated the branch as requested and all checks are passing.
I also addressed all the requested changes from @adamjstewart and left a comment summarizing them above.
Let me know if anything else is needed on my end. Thanks again!

adamjstewart commented on torchgeo/datamodules/solar_plants_brazil.py:
```diff
@@ -0,0 +1,104 @@
+from ..samplers.utils import _to_tuple
+from .geo import NonGeoDataModule
+
+# Per-channel statistics (mean and std) computed over the entire dataset.
```
Shouldn't they be computed only on the train split?

adamjstewart commented on torchgeo/datasets/solar_plants_brazil.py:
```diff
@@ -0,0 +1,176 @@
+        )
+
+    def __getitem__(self, index: int) -> dict[str, Tensor]:
+        """Retrieve an image-mask pair by index."""
```
Every function docstring should define inputs (Args) and outputs (Returns) other than self/None

adamjstewart commented on torchgeo/datasets/solar_plants_brazil.py:
```diff
@@ -0,0 +1,176 @@
+        bin_mask = (arr > 0).astype(np.uint8)
+        return torch.from_numpy(bin_mask).unsqueeze(0).long()
+
+    def plot(
```
Can you add a sample plot image to your PR description?

FederCO23 replied:
Added a sample output image from the `plot()` method to the PR description

FederCO23 commented on docs/api/datasets.rst:
```diff
@@ -492,6 +492,10 @@ So2Sat
 
 .. autoclass:: So2Sat
 
+SolarPlantsBrazil
```
Renamed to "Solar Plants Brazil" in datasets.rst as suggested

FederCO23 commented on tests/datamodules/test_solar_plants_brazil.py:
Removed `test_solar_plants_brazil.py`, added `tests/conf/solar_plants_brazil.yaml`, updated `test_segmentation.py`, and refreshed the dummy dataset with 256×256 instead of 32×32 images. Trainer + DataModule test runs successfully.

FederCO23 commented on tests/datasets/test_solar_plants_brazil.py:
```diff
@@ -0,0 +1,106 @@
+class TestSolarPlantsBrazil:
+    @pytest.fixture
+    def dataset_root(self) -> str:
+        # Point to the test data folder you generated and committed
```
Updated the return value to use `os.path.join` for cross-platform support. Also removed the comment, thanks for flagging it!

FederCO23 commented on tests/datasets/test_solar_plants_brazil.py:
```diff
@@ -0,0 +1,106 @@
+        assert torch.all(sample['image'] > 0)
+
+
+def test_download_called(monkeypatch: pytest.MonkeyPatch, tmp_path: Path) -> None:
```
Updated the `test_download_called` test to follow the same structure used in other dataset tests

FederCO23 commented on tests/datasets/test_solar_plants_brazil.py:
```diff
@@ -0,0 +1,106 @@
+    assert called['triggered']
+
+
+def test_missing_dataset_triggers_error(tmp_path: Path) -> None:
```
Moved the `test_missing_dataset_triggers_error` and `test_empty_split_folder_triggers_error` methods into the `TestSolarPlantsBrazil` class for consistency

FederCO23 commented on torchgeo/datamodules/solar_plants_brazil.py:
```diff
@@ -0,0 +1,104 @@
+            keepdim=True,
+        )
+
+    def setup(self, stage: str) -> None:
```
Removed the setup method since it’s redundant with the superclass. Cleaned up as suggested

FederCO23 commented on torchgeo/datamodules/solar_plants_brazil.py:
```diff
@@ -0,0 +1,104 @@
+            A batch of data.
+
+        """
+        # This solves a special case where if batch_size=1 the mask won't be stacked correctly
```
The dataset now returns mask with shape [H, W] instead of [1, H, W], aligning with the expected format for the loss function.
Unit tests and trainer test pass successfully

FederCO23 commented on torchgeo/datasets/solar_plants_brazil.py:
```diff
@@ -0,0 +1,176 @@
+    dataset provides satellite imagery and pixel-level annotations for detecting photovoltaic
+    solar power stations.
+
+    Dataset features:
```
Adjustments done. The docstring now renders correctly in the HTML

FederCO23 commented on torchgeo/datasets/solar_plants_brazil.py:
```diff
@@ -0,0 +1,176 @@
+    .. versionadded:: 0.8
+
+    Raises:
+        DatasetNotFoundError: If the dataset is not found and download=False.
```
Ok. It’s done

FederCO23 commented on torchgeo/datasets/solar_plants_brazil.py:
```diff
@@ -0,0 +1,176 @@
+    def __init__(
+        self,
+        root: Path = 'data',
+        split: str = 'train',
```
Replaced `split: str = 'train'` with `split: Literal['train', 'val', 'test'] = 'train'` to restrict valid values, as suggested. Thanks for pointing that out

FederCO23 commented:
@adamjstewart, thanks again for the detailed feedback!
All 16 requested changes have been addressed. I replied to each one directly above.
Let me know if there’s anything else needed on my end. Happy to contribute!

@robmarkcole, just tagging you here to keep you in the loop

adamjstewart commented on torchgeo/datasets/solar_plants_brazil.py:
```diff
@@ -0,0 +1,230 @@
+
+    """
+
+    url = 'https://huggingface.co/datasets/FederCO23/solar-plants-brazil/resolve/main/solarplantsbrazil.zip'
```
```suggestion
    url = 'https://huggingface.co/datasets/FederCO23/solar-plants-brazil/resolve/1dc13a453ef6acabf08a1781c523fd1db3d9bcc5/solarplantsbrazil.zip'
```
Use a stable commit hash so that the MD5 doesn't change even if the dataset is later updated.

adamjstewart commented on docs/api/datasets/non_geo_datasets.csv:
```diff
@@ -55,6 +55,7 @@ Dataset,Task,Source,License,# Samples,# Classes,Size (px),Resolution (m),Bands
 `SkyScript`_,IC,"NAIP, orthophotos, Planet SkySat, Sentinel-2, Landsat 8--9",MIT,5.2M,-,100--1000,0.1--30,RGB
 `So2Sat`_,C,Sentinel-1/2,"CC-BY-4.0","400,673",17,32x32,10,"SAR, MSI"
 `SODA`_,OD,Aerial,"CC-BY-NC-4.0","2513",9,"~2700x~4800","RGB"
+`Solar Plants Brazil`_,S,Aerial,"CC-BY-NC-4.0","272",2,256x256,10,"RGB + NIR"
```
License now matches on https://huggingface.co/datasets/FederCO23/solar-plants-brazil

adamjstewart commented on tests/datasets/test_solar_plants_brazil.py:
```diff
@@ -0,0 +1,100 @@
+        assert len(dataset) == 1
+
+    def test_plot(self, dataset: SolarPlantsBrazil) -> None:
+        sample = dataset[0].copy()
+        sample['prediction'] = sample['mask'].clone()
```
I don't think a copy or clone is actually needed is it?

FederCO23 replied:
You're right, since prediction is not being modified, both `.clone()` and `.copy()` are unnecessary and can be removed.
For reference, I had initially followed a similar pattern used in other datasets like `cloud_cover` and `chesapeake`, which use `.clone()` in their `test_plot()` methods."

adamjstewart commented on tests/datasets/test_solar_plants_brazil.py:
```diff
@@ -0,0 +1,100 @@
+        sample = dataset[0]
+        assert torch.all(sample['image'] > 0)
+
+    def test_download_called(
```
I don't think we test this for any other datasets, it's better to test that it's possible to "download" files from the `tests/data` directory to the test directory. See all of the other dataset tests.

FederCO23 replied:
Ok. Removed `test_download_called` and added `test_already_downloaded` and `test_not_downloaded` instead, following the structure used in other datasets.

adamjstewart commented on tests/datasets/test_solar_plants_brazil.py:
```diff
@@ -0,0 +1,100 @@
+
+        assert called['triggered']
+
+    def test_missing_dataset_triggers_error(self, tmp_path: Path) -> None:
```
Isn't this identical to `test_missing_dataset_raises`?

FederCO23 replied:
Removed `test_missing_dataset_triggers_error` (duplicate of `test_not_downloaded`) and renamed for consistency with other datasets

adamjstewart commented on torchgeo/datamodules/solar_plants_brazil.py:
```diff
@@ -0,0 +1,89 @@
+            batch_size: Number of samples per batch.
+            patch_size: Spatial dimensions (H, W) to crop from images.
+            num_workers: Number of subprocesses used to load the data.
+            **kwargs: Additional arguments passed to the dataset.
```
Passed to which dataset? We should link to the docs for that dataset like we do in all other datamodules

adamjstewart commented on torchgeo/datamodules/solar_plants_brazil.py:
```diff
@@ -0,0 +1,89 @@
+            num_workers: Number of subprocesses used to load the data.
+            **kwargs: Additional arguments passed to the dataset.
+
+        Returns:
+            None
```
This can be skipped if there is no Return value

adamjstewart commented on torchgeo/datamodules/solar_plants_brazil.py:
```diff
@@ -0,0 +1,89 @@
+            keepdim=True,
+        )
+
+    def on_after_batch_transfer(
```
This isn't doing anything and can be deleted

FederCO23 replied:
You're right, since no custom logic is added here. I've removed the `on_after_batch_transfer override`.

adamjstewart commented on torchgeo/datasets/solar_plants_brazil.py:
```diff
@@ -0,0 +1,230 @@
+
+    url = 'https://huggingface.co/datasets/FederCO23/solar-plants-brazil/resolve/main/solarplantsbrazil.zip'
+    bands = ('Red', 'Green', 'Blue', 'NIR')
+    md5 = 'DFA0D3EFDEF4143A33B0E7BA834EAAFA'
```
```suggestion
    md5 = 'dfa0d3efdef4143a33b0e7ba834eaafa'
```
Upper case doesn't work:
```pycon
> python3
Python 3.13.2 (main, Apr  8 2025, 15:27:33) [Clang 17.0.0 (clang-1700.0.13.3)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from torchgeo.datasets import SolarPlantsBrazil
>>> ds = SolarPlantsBrazil('data/solarplantsbrazil', download=True, checksum=True)
100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 134M/134M [00:06<00:00, 22.0MB/s]
Traceback (most recent call last):
  File "<python-input-1>", line 1, in <module>
    ds = SolarPlantsBrazil('data/solarplantsbrazil', download=True, checksum=True)
  File "/Users/Adam/Downloads/torchgeo/torchgeo/datasets/solar_plants_brazil.py", line 95, in __init__
    self._verify()
    ~~~~~~~~~~~~^^
  File "/Users/Adam/Downloads/torchgeo/torchgeo/datasets/solar_plants_brazil.py", line 123, in _verify
    self._download()
    ~~~~~~~~~~~~~~^^
  File "/Users/Adam/Downloads/torchgeo/torchgeo/datasets/solar_plants_brazil.py", line 131, in _download
    download_and_extract_archive(
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~^
        url=self.url,
        ^^^^^^^^^^^^^
    ...<2 lines>...
        md5=self.md5 if self.checksum else None,
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    )
    ^
  File "/Users/Adam/spack/var/spack/environments/default/.spack-env/view/lib/python3.13/site-packages/torchvision/datasets/utils.py", line 391, in download_and_extract_archive
    download_url(url, download_root, filename, md5)
    ~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/Adam/spack/var/spack/environments/default/.spack-env/view/lib/python3.13/site-packages/torchvision/datasets/utils.py", line 140, in download_url
    raise RuntimeError("File not found or corrupted.")
RuntimeError: File not found or corrupted.
```

adamjstewart commented on torchgeo/datasets/solar_plants_brazil.py:
```diff
@@ -0,0 +1,230 @@
+            None
+
+        """
+        if split not in ['train', 'val', 'test']:
+            raise ValueError(
+                f"Invalid split '{split}', expected one of: 'train', 'val', or 'test'"
+            )
```
Could honestly remove this and the test since we statically check that a valid Literal is used, but up to you. This is certainly nicer for users as it gives them a more clear explanation and is checked at runtime.

adamjstewart commented on torchgeo/datasets/solar_plants_brazil.py:
```diff
@@ -0,0 +1,230 @@
+    def _verify(self) -> None:
+        """Verify the dataset exists or download it.
+
+        Returns:
+            None
```
Skip

adamjstewart reviewed (changes requested):
Still quite a lot of differences between this dataset and the other builtin datasets

FederCO23 commented on torchgeo/datasets/solar_plants_brazil.py:
```diff
@@ -0,0 +1,230 @@
+        )
+
+        if len(self.image_paths) == 0:
+            raise DatasetNotFoundError(self)
```
I’ve now removed the `if len(self.image_paths) == 0` check from `__init__()` to align with the datasets pattern.
That said, I originally added this check to catch cases where the folder exists but is empty (ex: failed unzip). In that situation, `_verify()` would silently pass, but indexing into the dataset would fail later on in a less clear way.
I transfer that check to `_verify()`

FederCO23 commented on torchgeo/datasets/solar_plants_brazil.py:
```diff
@@ -0,0 +1,230 @@
+        """Retrieve an image-mask pair by index.
+
+        Args:
+            index (int): Index of the sample to retrieve.
```
Ok, this has been addressed

adamjstewart reviewed (changes requested):
Code looks much better now, thanks for all of the hard work!

Only remaining thing is to iron out the download logic and testing so we reach 100% coverage. The EuroSAT zipfile download logic is pretty clean and a good example:
```python
        directory = os.path.join(self.root, self.base_dir)                               
        zipfile = os.path.join(self.root, self.filename)                                 
        if os.path.isdir(directory):                                                     
            return                                                                       
        elif os.path.isfile(zipfile):                                                    
            extract_archive(zipfile)                                                     
        elif self.download:                                                              
            download_and_extract_archive(                                                
                self.url + self.filename,                                                
                self.root,                                                               
                md5=self.md5 if self.checksum else None,                                 
            )                                                                            
        else:                                                                            
            raise DatasetNotFoundError(self)
```
I actually don't like the EuroSAT tests that much. TreeSatAI has nice simple tests:
```python
    def test_download(self, dataset: TreeSatAI, tmp_path: Path, monkeypatch: MonkeyPatch) -> None:                
        monkeypatch.setattr(TreeSatAI, 'url', local_url) 
        TreeSatAI(tmp_path, download=True)                                               
                                                                                         
    def test_extract(self, dataset: TreeSatAI, tmp_path: Path) -> None:                  
        for file in glob.iglob(os.path.join(root, '*.*')):                               
            shutil.copy(file, tmp_path)                                                  
        TreeSatAI(tmp_path)                                                              
                                                                                         
    def test_not_downloaded(self, tmp_path: Path) -> None:                               
        with pytest.raises(DatasetNotFoundError, match='Dataset not found'):             
            TreeSatAI(tmp_path)    
```
Something simple like that should get you 100% test coverage.

FederCO23 commented on torchgeo/datasets/solar_plants_brazil.py:
```diff
@@ -0,0 +1,221 @@
+
+        self._verify()
+
+    def _verify(self) -> None:
```
I appreciate the clear guidance!
I’ve updated the `_verify()` method to handle the intermediate case where the zip file exists but hasn't been extracted yet, inspired by the EuroSAT logic as suggested, and also aligned the tests more closely with the TreeSatAI style.

Let me know if anything else needs attention. Thanks!

adamjstewart reviewed (approved):
Looks good now! @calebrob6 may be interested in this to help validate his global solar panel detection work.
